x-pack/plugins/infra/public/pages/metrics/metrics_explorer/hooks/use_metric_explorer_state.ts

```diff
@@ -27,49 +28,64 @@ export interface MetricExplorerViewState {
   id?: string;
 }
 
+const getTimestamps = (from: string, to: string) => {
+  const fromTimestamp = DateMath.parse(from)!.valueOf();
+  const toTimestamp = DateMath.parse(to, { roundUp: true })!.valueOf();
+
+  return {
+    fromTimestamp,
+    toTimestamp,
+  };
+};
+
 export const useMetricsExplorerState = (
   source: MetricsSourceConfigurationProperties,
   derivedIndexPattern: DataViewBase,
-  shouldLoadImmediately = true
+  enabled = true
 ) => {
-  const [refreshSignal, setRefreshSignal] = useState(0);
-  const [afterKey, setAfterKey] = useState<string | null | Record<string, string | null>>(null);
   const {
     defaultViewState,
     options,
-    currentTimerange,
+    timeRange,
     chartOptions,
     setChartOptions,
     setTimeRange,
     setOptions,
+    timestamps,
+    setTimestamps,
   } = useMetricsExplorerOptionsContainerContext();
 
-  const { loading, error, data, loadData } = useMetricsExplorerData(
+  const { data, error, fetchNextPage, isLoading } = useMetricsExplorerData(
     options,
     source,
     derivedIndexPattern,
-    currentTimerange,
-    afterKey,
-    refreshSignal,
-    shouldLoadImmediately
+    timestamps,
+    enabled
   );
 
-  const handleRefresh = useCallback(() => {
-    setAfterKey(null);
-    setRefreshSignal(refreshSignal + 1);
-  }, [refreshSignal]);
+  useEffect(() => {
+    setTimestamps({
+      interval: timeRange.interval,
+      ...getTimestamps(timeRange.from, timeRange.to),
+    });
```

Comment:
**suggestion:** as far as I can see, this effect replicates the same logic defined in the [`refetch` callback](https://github.com/elastic/kibana/pull/153045/files#diff-d18c7e63329a35adfb7b92292b7d49fea316686d77eec46cdaa0d307d0fa3d31R80) defined later on this file, you could move it above this effect and re-use it. I also see this has some unnecessary deps: `options, setOptions`.

Also, this effect is purely derivating the timestamp version from the inner `timeRange`, coming from the same hook, I wonder if we could move it inside the hook implementation, have you already seen any cons on this?

Reply:
I created `getTimestamps` to re-use it both in `refetch` and `useEffect`, I will give it a try to have the `setTimestamps` in a separate function to re-use in both places. It didn't seem to be necessary, but I don't oppose it either.

Regarding having `options, setOptions` in deps, that is intentional, since we need to recreate the timestamps whenever an option changes.

> I wonder if we could move it inside the hook implementation, have you already seen any cons on this?

If by hook you mean `useMetricsExplorerData`, I didn't keep that logic there because I wanted to have control over when we generate timestamps (when we change time range or options) and not when we run the hook. Still, if you think there is a way to make it work, I am open to suggestions

Reply:
I can clarify what I meant, it was not about extracting it into another function, but reuse what you already have created:
```js
const refetch = useCallback(() => {
  setTimestamps({
    interval: timeRange.interval,
    ...getTimestamps(timeRange.from, timeRange.to),
  });
}, [setTimestamps, timeRange]);

useEffect(() => {
  refetch()
}, [refetch, options, setOptions]);
```

Regarding the naming, `refetch` would represent a data fetching, but here it just updates timestamp values, shall this have a different name?

About `options, setOptions`as deps for the effect, it makes sense to use them in case you want to trigger the effect anyway on those changes, but would be nice to state the reason why you passed them in a comment for future readings of this code, anyone reading this snippet in 6 months could think those deps where not necessary and remove them :)

Reply:
Good point, I'll change `refetch` to `refresh`.

Reply:
Regarding `useCallback`, I will not add it unless it is necessary (There is an interesting article about it [here](https://www.developerway.com/posts/how-to-use-memo-use-callback))